### PR TITLE
[RHPAM-4608] Fix transaction isolation

### PR DESCRIPTION
**Pull request by ruromero**
Signed-off-by: ruromero <rromerom@redhat.com>

**JIRA**: 

[RHPAM-4608](https://issues.redhat.com/browse/RHPAM-4608)

There are 2 problems with transactions.

- TX starts for the whole migration and timeouts for long migrations. It has been isolated into individual JPA actions. i.e. saving reports and merging migration result
- Large migrations cannot be displayed in the UI due to big chunks of lobs not being able to be loaded. Now report logs are loaded independently and upon clicking on the (i) icon.

### CODE REVIEW

File: ddl-scripts/db2/db2-pim-schema.sql [hunk 1]
```diff
@@ -60,21 +60,16 @@ create table process_instance_ids (
...
+    foreign key (plan_id)
+    references plans;
 
 alter table process_instance_ids
-   add constraint FKobucfuy73fgsmkncl9q2rv6ko
-   foreign key (migration_definition_id)
-   references migrations;
+    add constraint FKobucfuy73fgsmkncl9q2rv6ko
+    foreign key (migration_definition_id)
+    references migrations;
```
Comment (afalhambra) [inline ddl-scripts/db2/db2-pim-schema.sql:75]: @ruromero  can you please remove unnecessary changes in all of the DDL scripts?

**Review by afalhambra — CHANGES_REQUESTED** [review]

File: ddl-scripts/db2/db2-pim-schema.sql (hunk 1 above)
Comment (ruromero) [reply ddl-scripts/db2/db2-pim-schema.sql:75]: I think these changes are necessary because formatting was not consistent

File: upgrade-scripts/postgresql/rhpam-7.13-to-7.14.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+alter table if exists migration_reports
```
Comment (afalhambra) [inline upgrade-scripts/postgresql/rhpam-7.13-to-7.14.sql]: There won't be 7.14 release, so I would just update current file https://github.com/kiegroup/process-migration-service/blob/main/upgrade-scripts/postgresql/rhpam-7.12-to-7.13.sql to include these changes when migrating to 7.13 version

File: upgrade-scripts/postgresql/pim-7.73-to-7.74.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+alter table if exists migration_reports
```
Comment (afalhambra) [inline upgrade-scripts/postgresql/pim-7.73-to-7.74.sql]: right name should be `pim-7.71-to-7.73.sql` (there won't be any 7.74)

**Review by afalhambra — CHANGES_REQUESTED** [review]

File: upgrade-scripts/postgresql/rhpam-7.13-to-7.14.sql (hunk 2 above)
Comment (ruromero) [reply upgrade-scripts/postgresql/rhpam-7.13-to-7.14.sql]: Done

File: upgrade-scripts/postgresql/pim-7.73-to-7.74.sql (hunk 3 above)
Comment (ruromero) [reply upgrade-scripts/postgresql/pim-7.73-to-7.74.sql]: Done

File: ddl-scripts/db2/db2-pim-schema.sql (hunk 1 above)
Comment (afalhambra) [reply ddl-scripts/db2/db2-pim-schema.sql:75]: ok, thanks

File: src/test/resources/initial-ddls/db2/db2-pim-schema.sql [hunk 4]
```diff
@@ -60,21 +60,16 @@ create table process_instance_ids (
 create index IDX_MigrationReports_Id on migration_reports (migration_id);
 
 alter table migration_report_logs
-   add constraint FKj8bsydiucvs2kygnscp1bt1wy
```
Comment (afalhambra) [inline src/test/resources/initial-ddls/db2/db2-pim-schema.sql]: all the DDL scripts located under initial-ddls folder shouldn't be modified. This folder contains the original/initial DDL scripts created, and they are used to verify/test the upgrade process through the rest of the upgrade DDL scripts.

File: src/main/java/org/kie/processmigration/model/MigrationReportDto.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
```
Comment (afalhambra) [inline src/main/java/org/kie/processmigration/model/MigrationReportDto.java]:
```suggestion
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```

File: src/main/java/org/kie/processmigration/model/MigrationReportDto.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import java.time.Instant;
+import java.util.List;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import lombok.experimental.Accessors;
+
+@EqualsAndHashCode(callSuper = false)
```
Comment (afalhambra) [inline src/main/java/org/kie/processmigration/model/MigrationReportDto.java]: Not needed. Default is set to `false`

File: src/main/java/org/kie/processmigration/model/Migration.java [hunk 7]
```diff
@@ -50,7 +43,7 @@
 
 @Entity
 @Table(name = "migrations")
-@EqualsAndHashCode
+@EqualsAndHashCode(callSuper = false)
```
Comment (afalhambra) [inline src/main/java/org/kie/processmigration/model/Migration.java:46]: Not needed. Default is set to `false`

File: src/main/java/org/kie/processmigration/rest/MigrationResource.java [hunk 8]
```diff
@@ -73,6 +73,12 @@ public Response getResults(@PathParam("id") Long id) throws MigrationNotFoundExc
         return Response.ok(migrationService.getResults(id)).build();
     }
 
+    @GET
+    @Path("/{id}/results/{logId}")
+    public Response getResultLog(@PathParam("id") Long id, @PathParam("logId") Long logId) throws MigrationNotFoundException {
```
Comment (afalhambra) [inline src/main/java/org/kie/processmigration/rest/MigrationResource.java]: Not thrown. Please remove
```suggestion
    public Response getResultLog(@PathParam("id") Long id, @PathParam("logId") Long logId) {
```

File: src/main/java/org/kie/processmigration/service/impl/PlanServiceImpl.java [hunk 9]
```diff
@@ -30,10 +30,11 @@ public class PlanServiceImpl implements PlanService {
 
     @Override
     public List<Plan> findAll() {
```
Comment (afalhambra) [inline src/main/java/org/kie/processmigration/service/impl/PlanServiceImpl.java:33]: Shouldn't this be transactional as well? wdyt?

File: ddl-scripts/postgresql-plus/postgresql-plus-pim-drop-schema.sql [hunk 10]
```diff
@@ -0,0 +1,19 @@
+alter table migration_report_logs
```
Comment (afalhambra) [inline ddl-scripts/postgresql-plus/postgresql-plus-pim-drop-schema.sql]: Please remove the files related to Postgres Plus.

File: ddl-scripts/postgresql-plus/postgresql-plus-pim-schema.sql [hunk 11]
```diff
@@ -0,0 +1,75 @@
+create sequence MIG_REP_ID_SEQ start 1 increment 50;
```
Comment (afalhambra) [inline ddl-scripts/postgresql-plus/postgresql-plus-pim-schema.sql]: Please remove the files related to Postgres Plus.

File: src/test/resources/initial-ddls/postgresql-plus/postgresql-plus-pim-drop-schema.sql [hunk 12]
```diff
@@ -0,0 +1,19 @@
+alter table migration_report_logs
```
Comment (afalhambra) [inline src/test/resources/initial-ddls/postgresql-plus/postgresql-plus-pim-drop-schema.sql]: Please remove the files related to Postgres Plus.

File: src/test/resources/initial-ddls/postgresql-plus/postgresql-plus-pim-schema.sql [hunk 13]
```diff
@@ -0,0 +1,75 @@
+create sequence MIG_REP_ID_SEQ start 1 increment 50;
```
Comment (afalhambra) [inline src/test/resources/initial-ddls/postgresql-plus/postgresql-plus-pim-schema.sql]: Please remove the files related to Postgres Plus.

File: src/test/java/org/kie/processmigration/integration/ProcessMigrationIT.java [hunk 14]
```diff
@@ -20,6 +20,7 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.util.ArrayList;
```
Comment (afalhambra) [inline src/test/java/org/kie/processmigration/integration/ProcessMigrationIT.java]: not needed

**Review by afalhambra — CHANGES_REQUESTED** [review]

File: src/main/java/org/kie/processmigration/model/Migration.java (hunk 7 above)
Comment (ruromero) [reply src/main/java/org/kie/processmigration/model/Migration.java:46]: Compiler asks for this to explicitly set to true/false when the class doesn't extend Object

```
Generating equals/hashCode implementation but without a call to superclass, even though this class does not extend java.lang.Object. If this is intentional, add '@EqualsAndHashCode(callSuper=false)'
```

**Review by gmunozfe — COMMENTED** [review]
Some methods in `MigrationServiceImpl` has no `@Transactional` annotation but use transactions as `submit` and `migrate`. Do you think they are needed?

**Comment by ruromero** [comment]
> Some methods in `MigrationServiceImpl` has no `@Transactional` annotation but use transactions as `submit` and `migrate`. Do you think they are needed?

Yes, the migrate will not start a transaction but small transactions for persisting the reports and the result.
